### ISSUE TITLE
extract_image_patches

### ISSUE THREAD

**Issue by iperov** [plaidml/plaidml#243]
how to implement differentiable extract_image_patches in plaidML same or similar to
[tf.image.extract_image_patches](https://www.tensorflow.org/api_docs/python/tf/image/extract_image_patches)?
tf version is differentiable

**Comment by kilroythethird** [comment]
+1 for a plaidml implementation of extract_image_patches or at least a workaround.
Required fe. to implement DSSIM Loss Function.

**Comment by iperov** [comment]
@kilroythethird I already implemented tensorflow SSIM on pure keras that works on plaidML.

```
def dssim(kernel_size=11, k1=0.01, k2=0.03, max_value=1.0):
    # port of tf.image.ssim to pure keras in order to work on plaidML backend.

    def func(y_true, y_pred):
        ch = K.shape(y_pred)[-1]

        def _fspecial_gauss(size, sigma):
            #Function to mimic the 'fspecial' gaussian MATLAB function.
            coords = np.arange(0, size, dtype=K.floatx())
            coords -= (size - 1 ) / 2.0
            g = coords**2
            g *= ( -0.5 / (sigma**2) )
            g = np.reshape (g, (1,-1)) + np.reshape(g, (-1,1) )
            g = K.constant ( np.reshape (g, (1,-1)) )
            g = K.softmax(g)
            g = K.reshape (g, (size, size, 1, 1)) 
            g = K.tile (g, (1,1,ch,1))
            return g
                      
        kernel = _fspecial_gauss(kernel_size,1.5)

        def reducer(x):
            return K.depthwise_conv2d(x, kernel, strides=(1, 1), padding='valid')

        c1 = (k1 * max_value) ** 2
        c2 = (k2 * max_value) ** 2
        
        mean0 = reducer(y_true)
        mean1 = reducer(y_pred)
        num0 = mean0 * mean1 * 2.0
        den0 = K.square(mean0) + K.square(mean1)
        luminance = (num0 + c1) / (den0 + c1)
        
        num1 = reducer(y_true * y_pred) * 2.0
        den1 = reducer(K.square(y_true) + K.square(y_pred))
        c2 *= 1.0 #compensation factor
        cs = (num1 - num0 + c2) / (den1 - den0 + c2)

        ssim_val = K.mean(luminance * cs, axis=(-3, -2) )
        return K.mean( (1.0 - ssim_val ) / 2.0 )

    return func
```
usage:
loss = dssim()( y_true, y_pred )

**Comment by kilroythethird** [comment]
Awesome. Thanks a lot for sharing.

**Comment by iperov** [comment]
but anyway I need a SUBJ :)
I cannot understand how to use such math language like Tile.
Impossibru to understand for non-math man like me :rofl:

**Comment by brianretford** [comment]
Hey Tim, can we Tile-ify this? Alternatively we'd accept the patch but I bet peformance isn't great with out an optimizing graph frontend like ngraph

**Comment by brianretford** [comment]
Also just FYI, we're hiring: https://jobs.intel.com/ShowJob/Id/1753461/Senior-Tensor-Compiler-Engineer/

We have multiple open reqs that aren't public quite yet at which point they'll all be posted to the front page

**Comment by SleepProgger** [comment]
Definitely not PR worth, but i wrote an extract_image_patches fallback.
It doesn't support strides != ksize and rates != (1,1,1,1). Also it probably is horrible code, but thats as far as i got before my brain melted ;)

```
class ImagePatches(plaidml.tile.Operation):
    def __init__(self, images, ksizes, strides=None, rates=None, padding="VALID"):
        """
        Should be compatible to tf.extract_image_patches
        Does NOT support rates != (1,1,1,1) and strides != ksizes.
        """
        if strides is not None and tuple(strides) != tuple(ksizes):
            raise NotImplementedError("Strides not supported")
        if rates is not None and tuple(rates) != (1,1,1,1):
            raise NotImplementedError("Rates not supported")
        code = """
            function (I[B, Y, X, D]) -> (O) {{
                TMP[b, ny, nx, y, x, d: B, Y / {YS}, X / {XS}, {YS}, {XS}, D] =
                    =(I[b, ny * {YS} + y, nx * {XS} + x, d]);
                O = reshape(TMP, B, Y / {YS}, X / {XS}, D * {YS} * {XS});
            }}
        """.format(YS=ksizes[1], XS=ksizes[2])
        o_shape = list(images.shape.dims)
        if padding.upper() != "VALID":
            pad_y = o_shape[1]%ksizes[1]
            pad_x = o_shape[2]%ksizes[2]
            images = K.spatial_2d_padding(images, ((0, pad_y), (0, pad_x)) )
            o_shape = list(images.shape.dims)
        o_shape[1] //= ksizes[1]
        o_shape[2] //= ksizes[2]
        o_shape[3] *= ksizes[1]*ksizes[2]
        super(ImagePatches, self).__init__(code,
                [('I', images),],
                [('O', plaidml.tile.Shape(images.shape.dtype, o_shape))]
        )
```

Maybe it helps someone.

**Comment by SleepProgger** [comment]
Since #303 merged an extract_image_patches version as `op.extract_image_patches` this can be closed, no ?

**Comment by brianretford** [comment]
Yeah! We are going to do a bug scrub later this week. We'll get your other patches merged in and release 6.1 soon, now that I have a dedicated person to work on this stuff.
